src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -0,0 +1,237 @@
+	 * recreated.
+	 */
+	if (strcmp(schemaName, "public") == 0)
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * ShouldCreateTenantTable returns true if given parsetree is a CREATE TABLE
+ * statement and the table should be treated as a tenant table.
+ */
+bool
+ShouldCreateTenantTable(Oid relationId)
+{
+	if (IsBinaryUpgrade)
+	{
+		return false;
+	}
+
+	/*
+	 * CREATE TABLE commands issued by internal backends are not meant to
+	 * create tenant tables but to sync metadata.
+	 */
+	if (IsCitusInternalBackend() || IsRebalancerInternalBackend())
+	{
+		return false;
+	}
+
+	Oid schemaId = get_rel_namespace(relationId);
+	return IsTenantSchema(schemaId);
+}
+
+
+/*
+ * CreateTenantTable creates a tenant table with given relationId.
+ *
+ * This means creating a single shard distributed table without a shard
+ * key and colocating it with the other tables in its schema.
+ */
+void
+CreateTenantTable(Oid relationId)
+{
+	if (!IsCoordinator())
+	{
+		/*
+		 * We don't support creating tenant tables from workers. We could
+		 * let ShouldCreateTenantTable() to return false to allow users to
+		 * create a local table as usual but that would be confusing because
+		 * it might sound like we allow creating tenant tables from workers.
+		 * For this reason, we prefer to throw an error instead.
+		 *
+		 * Indeed, CreateSingleShardTable() would already do so but we
+		 * prefer to throw an error with a more meaningful message, rather
+		 * than saying "operation is not allowed on this node".
+		 */
+		ereport(ERROR, (errmsg("cannot create a tenant table from a worker node"),
+						errhint("Connect to the coordinator node and try again.")));
+	}
+
+	/*
+	 * Decide name of the table with lowest oid in the colocation group
+	 * and use it as the colocate_with parameter.
+	 */
+	char *colocateWithTableName = "none";
+
+	/*
+	 * Acquire default colocation lock to prevent concurrently forming
+	 * multiple colocation groups for the same schema.
+	 *
+	 * Note that the lock is based on schemaId to avoid serializing
+	 * default colocation group creation for all tenant schemas.
+	 */
+	Oid schemaId = get_rel_namespace(relationId);
+	AcquireCitusTenantSchemaDefaultColocationLock(schemaId);
+
+	uint32 colocationId = SchemaIdGetTenantColocationId(schemaId);
+	bool firstTableInSchema = (colocationId == INVALID_COLOCATION_ID);
+	if (!firstTableInSchema)
+	{
+		/*
+		 * Release the lock if the schema is already associated with a
+		 * colocation group.
+		 */
+		ReleaseCitusTenantSchemaDefaultColocationLock(schemaId);
+
+		Oid colocateWithTableId = ColocationGroupGetTableWithLowestOid(colocationId);
+		colocateWithTableName = generate_qualified_relation_name(colocateWithTableId);
+	}
+
+	CreateSingleShardTable(relationId, colocateWithTableName);
+
+	if (firstTableInSchema)
+	{
+		/*
+		 * Save it into pg_dist_tenant_schema if this is the first tenant
+		 * table in the schema.
+		 */
+		SetTenantSchemaColocationId(schemaId, TableColocationId(relationId));
```

Comment:
I have some concerns about the approach of repeatedly resetting and recreating the colocation ID when a schema goes empty. As a user, I expect a strong coupling between the schema and my tenant, but this effectively creates a new tenant for the purpose of citus_stat_tenants and future tenant pinning. Is there a way to preserve the co-location ID of empty schemas?